hedera-node/src/main/java/com/hedera/services/txns/contract/helpers/StorageExpiry.java

```diff
@@ -0,0 +1,150 @@
+ * Helper to efficiently get the expiration time for new storage being allocated in a {@link MessageFrame}.
+ *
+ * <p>Note the allocating contract is given by {@link MessageFrame#getRecipientAddress()}. Now there are
+ * two cases: either the contract allocating the storage <i>already existed</i> at the start of the EVM
+ * transaction; or we have created the allocating contract in the ongoing EVM transaction.
+ *
+ * <ol>
+ *     <li>In the first case, the storage expiry is just the same as the (well-known) expiry of
+ *     the allocating contract---which must be available in the accounts {@code MerkleMap}.</li>
+ *     <li>In the second case, the expiry of the newly-created, allocating contract is determined by its
+ *     "sponsor chain" (c.f. {@link HederaWorldState#customizeSponsoredAccounts()}).
+ * </ol>
+ * Note we say "sponsor <i>chain</i>"---and not simply "sponsor"---because the sponsor of the allocating
+ * contract might have itself been created in the ongoing EVM transaction! So we have to walk up the
+ * {@code MessageFrame} stack until we find a {@code recipient} contract that already existed; this will
+ * be the source of expiries for the whole sponsor chain.
+ *
+ * <p><b>IMPORTANT:</b> if our stack walk never finds a {@code recipient} that already exists, the
+ * of the sponsor chain must begin with a contract that is <i>itself</i> being created this EVM
+ * transaction. In this case, the HAPI {@code cryptoCreate} should be the ultimate source of the
+ * storage expiry. So we need to option to configure the oracle with a {@code fallbackExpiryFromHapi}.
+ */
+@Singleton
+public class StorageExpiry {
+	private static final Logger log = LogManager.getLogger(StorageExpiry.class);
+
+	private static final long UNAVAILABLE_EXPIRY = 0;
+
+	private final Oracle callOracle = new Oracle(UNAVAILABLE_EXPIRY);
+	private final UnusableStaticOracle unusableStaticOracle = new UnusableStaticOracle();
+
+	private final AliasManager aliasManager;
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts;
+
+	@Inject
+	public StorageExpiry(
+			final AliasManager aliasManager,
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts
+	) {
+		this.aliasManager = aliasManager;
+		this.contracts = contracts;
+	}
+
+	public Oracle hapiCreationOracle(final long hapiExpiry) {
+		return new Oracle(hapiExpiry);
+	}
+
+	public Oracle hapiCallOracle() {
+		return callOracle;
+	}
+
+	public Oracle hapiStaticCallOracle() {
+		return unusableStaticOracle;
+	}
+
+	public class Oracle {
+		private final long fallbackExpiry;
+
+		private Oracle(long fallbackExpiryFromHapi) {
+			this.fallbackExpiry = fallbackExpiryFromHapi;
+		}
+
+		/**
+		 * Returns the effective expiry for storage being allocated in the current frame.
+		 *
+		 * @param frame the active message frame
+		 * @return the effective expiry for allocated storage
+		 */
+		public long storageExpiryIn(final MessageFrame frame) {
+			final var curContracts = contracts.get();
```

Comment:
If the contract is being created in this transaction, we won't have set its `EXPIRY` in the ledger here (since that happens after `commit()` in `customizeSponsoredAccounts()`).

So for this purpose, it's better to look at just the contracts that existed at the start of the transaction. 👍